lib/streamlit/elements/lib/subtitle_utils.py

```diff
@@ -0,0 +1,164 @@
+        The content of the .srt file as a string or a bytes stream.
+    Returns
+    -------
+    bytes
+        The content converted into .vtt format.
+    """
+
+    # If the input is a bytes stream, convert it to a string
+    if isinstance(srt_data, bytes):
+        # Decode the bytes to a UTF-8 string
+        try:
+            srt_data = srt_data.decode("utf-8")
+        except UnicodeDecodeError as e:
+            raise ValueError("Could not decode the input stream as UTF-8.") from e
+    if not isinstance(srt_data, str):
+        # If it's not a string by this point, something is wrong.
+        raise TypeError(
+            f"Input must be a string or a bytes stream, not {type(srt_data)}."
+        )
+
+    # Replace SubRip timing with WebVTT timing
+    vtt_data = re.sub(r"(\d{2}:\d{2}:\d{2}),(\d{3})", r"\1.\2", srt_data)
+
+    # Add WebVTT file header
+    vtt_content = "WEBVTT\n\n" + vtt_data
+    # Convert the vtt content to bytes
+    vtt_content = vtt_content.strip().encode("utf-8")
+
+    return vtt_content
+
+
+def process_subtitle_data(
+    coordinates: str,
+    data: str | bytes | io.BytesIO,
+    label: str,
+) -> str:
+    allowed_formats = {".srt", ".vtt"}
+
+    def handle_string_data(data_str: str) -> bytes:
+        """Handles string data, either as a file path or raw content."""
+        if os.path.isfile(data_str):
+            path = Path(data_str)
+            file_extension = path.suffix.lower()
+
+            if file_extension not in allowed_formats:
+                raise ValueError(
+                    f"Incorrect subtitle format {file_extension}. Subtitles must be in "
+                    f"one of the following formats: {', '.join(allowed_formats)}"
+                )
+            with open(data_str, "rb") as file:
+                content = file.read()
+            return srt_to_vtt(content) if file_extension == ".srt" else content
+
+        content_string = data_str.strip()
+
+        if content_string.startswith("WEBVTT") or content_string == "":
+            return content_string.encode("utf-8")
+        elif _is_srt(content_string):
+            return srt_to_vtt(content_string)
+        raise ValueError(
+            "The provided string neither matches valid VTT nor SRT format."
+        )
+
+    def handle_stream_data(stream: io.BytesIO) -> bytes:
+        """Handles io.BytesIO data, assuming it's SRT content."""
+        stream.seek(0)
+        stream_data = stream.getvalue()
+        return srt_to_vtt(stream_data) if _is_srt(stream) else stream_data
+
+    # Determine the type of data and process accordingly
+    if isinstance(data, str):
+        subtitle_data = handle_string_data(data)
+    elif isinstance(data, bytes):
+        subtitle_data = data  # Assume bytes are already in the correct format.
+    elif isinstance(data, io.BytesIO):
+        subtitle_data = handle_stream_data(data)
+    else:
+        raise RuntimeError(f"Invalid binary data format for subtitle: {type(data)}.")
```

Comment:
Good catch. We should raise a RuntimeError only when it doesn't follow into the other error categories. Let's raise the appropriate error here if you know which is most suitable. Probably a `TypeError` because it didn't correspond to any of the types in the if/else checks, right?

Reply:
yeah, let's do `TypeError`, anyway we catch it on upper level to raise an `StreamlitApiException`, so we will have nice tracebacks.